### PR TITLE
Replace LinkedObjectPool with ArrayObjectPool for better performance

### LINKED ISSUE #15: Introduce Lower Latency More Deterministic ArrayObjectPool
Introduce an `ArrayObjectPool` to achieve ultra-low latency performance suitable for high-performance, single-threaded systems.

- **Implementation Highlights:**
  -  **Preallocated Array:** Uses a preallocated array to improve cache locality and eliminate runtime allocations during normal operation.
  - **Instance Management:** Objects are reused efficiently, and new instances are created on-demand when the pool is empty.
  - **Discarding Policy:** When the pool is full, fallback to using LinkedObjectPool.
- **Testing and Validation:**
  - **Behavioral Tests:** Validated that the pool behaves as expected, especially in edge cases like running out of instances or attempting to release objects when the ArrayObjectPool pool is full.
- **Documentation:** Add JavaDocs and inline comments to explain the rationale behind design decisions and to provide guidance on usage and potential pitfalls.

Replace all uses of LinkedObjectPool with ArrayObjectPool.

### PR DESCRIPTION
**Pull request by dakotahNorth**
### **User description**
Replace LinkedObjectPool with ArrayObjectPool for better performance. Updated NoGCTest for improved metrics and removed unused GitHub Maven CI workflow.

___

### **PR Type**
enhancement, tests, other

___

### **Description**
- Replaced `LinkedObjectPool` with `ArrayObjectPool` in `OrderBook` for better performance and increased pool sizes.
- Introduced `ArrayObjectPool` class designed for ultra-low latency and single-threaded systems.
- Enhanced `NoGCTest` to include performance metrics and increased garbage creation for testing.
- Refactored `LinkedObjectPool` to use `Supplier` instead of `Builder` for object creation.
- Optimized `SystemTimestamper` to use `System.nanoTime()` for better precision.
- Removed unused GitHub Maven CI workflow.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>OrderBook.java</strong><dd><code>Use ArrayObjectPool for Order and PriceLevel</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/main/java/com/coralblocks/coralme/OrderBook.java

<li>Replaced <code>LinkedObjectPool</code> with <code>ArrayObjectPool</code> for <code>Order</code> and <br><code>PriceLevel</code>.<br> <li> Increased pool size from 8 to 32.<br>

</details>

  </td>
  <td><a href="https://github.com/dakotahNorth/CoralME/pull/16/files#diff-0b8b5d4032c2d6c29f44f01b947259e455314f5869344a7691e6e3b2d1e6873d">+3/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>ArrayObjectPool.java</strong><dd><code>Introduce ArrayObjectPool for efficient object pooling</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/main/java/com/coralblocks/coralme/util/ArrayObjectPool.java

<li>Introduced new <code>ArrayObjectPool</code> class for object pooling.<br> <li> Designed for single-threaded systems with ultra-low latency.<br>

</details>

  </td>
  <td><a href="https://github.com/dakotahNorth/CoralME/pull/16/files#diff-8a1c81f4106862c2feb12c35720adaf9058ac5fc354a75180b24468d0aa04690">+123/-0</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>LinkedObjectPool.java</strong><dd><code>Refactor LinkedObjectPool to use Supplier</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/main/java/com/coralblocks/coralme/util/LinkedObjectPool.java

<li>Replaced <code>Builder</code> with <code>Supplier</code> for object creation.<br> <li> Improved exception handling in constructor.<br>

</details>

  </td>
  <td><a href="https://github.com/dakotahNorth/CoralME/pull/16/files#diff-36619ba7455262faa74ce70e6fa0d854da9df98c5f172bfb849fc78b70ed507f">+34/-33</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>SystemTimestamper.java</strong><dd><code>Optimize SystemTimestamper for nanosecond precision</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/main/java/com/coralblocks/coralme/util/SystemTimestamper.java

<li>Changed <code>nanoEpoch</code> method to use <code>System.nanoTime()</code>.<br> <li> Made class <code>final</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/dakotahNorth/CoralME/pull/16/files#diff-c6d2375ae0f9c2337a86eb1cad6bd37a75b5ed3940efa5d94a67071d3376a971">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>NoGCTest.java</strong><dd><code>Enhance NoGCTest with performance metrics</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/main/java/com/coralblocks/coralme/example/NoGCTest.java

<li>Updated command-line example formatting.<br> <li> Added timing metrics for book operations.<br> <li> Increased garbage creation in test loop.<br>

</details>

  </td>
  <td><a href="https://github.com/dakotahNorth/CoralME/pull/16/files#diff-9e369b6add939752d6bb1d0eb6fd0d3f6ee7420eea429fdc3fc2bbfce336b580">+16/-5</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Other</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>maven.yml</strong><dd><code>Remove unused GitHub Maven CI workflow</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.github/workflows/maven.yml

- Removed unused GitHub Maven CI workflow.

</details>

  </td>
  <td><a href="https://github.com/dakotahNorth/CoralME/pull/16/files#diff-5dbf1a803ecc13ff945a08ed3eb09149a83615e83f15320550af8e3a91976446">+0/-35</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

File: src/main/java/com/coralblocks/coralme/util/SystemTimestamper.java [hunk 1]
```diff
@@ -30,6 +30,6 @@ public class SystemTimestamper implements Timestamper {
 	 */
 	@Override
 	public long nanoEpoch() {
-		return System.currentTimeMillis() * 1000000L;
+		return System.nanoTime();
```
[the remark this answers is not in the document]
Comment (dakotahNorth) [reply src/main/java/com/coralblocks/coralme/util/SystemTimestamper.java]: Updated the calculation to return true time.

File: src/main/java/com/coralblocks/coralme/util/ArrayObjectPool.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+ * // Use the object
+ * // ...
+ *
+ * // Release the object back to the pool
+ * objectPool.release(obj);
+ * }</pre>
+ *
+ * @param <E> the type of objects this object pool will hold
+ */
+public final class ArrayObjectPool<E> implements ObjectPool<E> {
```
[the remark this answers is not in the document]
Comment (dakotahNorth) [reply src/main/java/com/coralblocks/coralme/util/ArrayObjectPool.java:53]: For ultra-low latency Java apps, AutoCloseable is generally not the best fit unless there is a well-defined scenario where the object pool needs to be shut down and cleaned up deterministically. If resource cleanup must be handled, it might be better to manage it externally in a controlled shutdown process, ensuring that such cleanup doesn’t interfere with critical low-latency paths.
